src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::{Either, Itertools};
+use risingwave_common::types::DataType;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+use risingwave_expr::aggregate::AggKind;
+
+use super::super::plan_node::*;
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, FunctionCall, InputRef};
+use crate::optimizer::plan_node::generic::{Agg, GenericPlanNode, GenericPlanRef};
+use crate::optimizer::plan_visitor::{PlanCorrelatedIdFinder, PlanVisitor};
+use crate::optimizer::PlanRef;
+use crate::utils::{Condition, IndexSet};
+
+/// Pull up correlated predicates from the right agg side of Apply to the `on` clause of Join.
+///
+/// Before:
+///
+/// ```text
+///     LogicalApply
+///    /            \
+///  LHS          Project
+///                 |
+///                Agg [group by nothing]
+///                 |
+///               Project
+///                 |
+///               Filter [correlated_input_ref(yyy) = xxx]
+/// ```
+///
+/// After:
+///
+/// ```text
+///     LogicalApply [yyy = xxx]
+///    /            \
+///  LHS          Project
+///                 |
+///                Agg [group by xxx]
+///                 |
+///               Project
+///                 |
+///               Filter
+/// ```
```

Comment:
A graph to explain how this rule works. It tries to pull up the correlated expr from the filter to the apply.